### PR TITLE
Deprecate adding two Time instances with ActiveSupport::TimeWithZone

### PR DESCRIPTION
**Pull request by Schwad**
Ref: #52084

Subtracting two time instances results in a Float representing the time between the instances. This is correct behaviour.

Adding two time instances should not work, but it does. It is an easily made mistake.

```ruby
10.days.ago + 10.days.ago

# => Mon, 09 Jan 2079 21:03:46.179229000 UTC +00:0
```

 [This change](https://github.com/rails/rails/pull/52084/files#diff-aa0ae5ccf92f812f874b632afe70375c52772636f927fe6e34ffeaebf54af9d1L303) removed the rescue statement that had made this possible.

Now, it is a breaking change to have made that easy mistake. This PR instead deprecates this change until Rails 8.0 and raises a deprecation warning when adding two Time instances to give people time to fix their code.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

File: activesupport/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,9 @@
+*   Deprecate adding two Time instances with ActiveSupport::TimeWithZone
+    Right now adding two time instances together such as `10.days.ago + 10.days.ago` produces a nonsensical future date.
+    Includes a deprecation to remove this behaviour in Rails 8.0
+
+    *Nick Schwaderer*
```
Comment (eileencodes) [inline activesupport/CHANGELOG.md:5]: I think this reads better and is clearer like this:

```suggestion
*   Deprecate addition for `Time` instances with `ActiveSupport::TimeWithZone`.

    Currently adding time instances together such as `10.days.ago + 10.days.ago` produces a nonsensical future date. This behavior is deprecated and will be removed in Rails 8.0.

    *Nick Schwaderer*
```

File: activesupport/lib/active_support/time_with_zone.rb [hunk 2]
```diff
@@ -303,6 +303,13 @@ def +(other)
         result = utc + other
         result.in_time_zone(time_zone)
       end
+    rescue TypeError
+      ActiveSupport.deprecator.warn(
+        "Do not add an instance of #{other.class} to an instance of #{utc.class}. This behaviour will raise " \
+        "in Rails 8.0."
```
Comment (eileencodes) [inline activesupport/lib/active_support/time_with_zone.rb:309]: We've standardized Rails on US english. Generally deprecations are also written as what is deprecated, rather than telling someone not to do something.

```suggestion
        "Adding an instance of #{other.class} to an instance of #{utc.class} is deprecated. This behavior will raise " \
        "a `TypeError` in Rails 8.0."
```

**Review by eileencodes — CHANGES_REQUESTED** [review]
I think deprecating this behavior is correct since a lot of apps could be depending on it. I've left some suggestions to improve the language around the deprecation and in the changelog to better match Rails voice style.

**Comment by Schwad** [comment]
@eileencodes thank you very much for the guidance around the Rails voice style for the deprecation && CHANGELOG. 🙇 That's much better.

**Comment by eileencodes** [comment]
Thanks @Schwad - can you squash your commits so there's only one with the commit message about what's changing/why?

File: activesupport/lib/active_support/time_with_zone.rb [hunk 3]
```diff
@@ -537,6 +544,7 @@ def respond_to?(sym, include_priv = false)
     # Ensure proxy class responds to all methods that underlying time instance
     # responds to.
     def respond_to_missing?(sym, include_priv)
+      return false if sym.to_sym == :acts_like_date?
```
Comment (jhawthorn) [inline activesupport/lib/active_support/time_with_zone.rb:547]: Is this change necessary? This should still be impossible.

File: activesupport/lib/active_support/time_with_zone.rb [hunk 4]
```diff
@@ -303,6 +303,13 @@ def +(other)
         result = utc + other
         result.in_time_zone(time_zone)
       end
+    rescue TypeError
+      ActiveSupport.deprecator.warn(
+        "Adding an instance of #{other.class} to an instance of #{utc.class} is deprecated. This behavior will raise " \
+        "a `TypeError` in Rails 8.0."
+      )
+      result = utc.acts_like?(:date) ? utc.since(other) : utc + other rescue utc.since(other)
```
Comment (jhawthorn) [inline activesupport/lib/active_support/time_with_zone.rb:311]: As far as I know `utc.acts_like?(:date)` should always be false, and we've already tried utc + other and know it raises a TypeError.

Can this be ...?

```suggestion
      result = utc.since(other)
```

File: activesupport/CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,9 @@
+*   Deprecate addition for `Time` instances with `ActiveSupport::TimeWithZone`.
+
+    Currently adding time instances together such as `10.days.ago + 10.days.ago` produces a nonsensical future date. This behavior is deprecated and will be removed in Rails 8.0.
```
Comment (jhawthorn) [inline activesupport/CHANGELOG.md:3]: (non-blocking)

```suggestion
    Previously adding time instances together such as `10.days.ago + 10.days.ago` produced a nonsensical future date. This behavior is deprecated and will be removed in Rails 8.0.
```

File: activesupport/lib/active_support/time_with_zone.rb [hunk 6]
```diff
@@ -303,6 +303,13 @@ def +(other)
         result = utc + other
         result.in_time_zone(time_zone)
       end
+    rescue TypeError
+      ActiveSupport.deprecator.warn(
+        "Adding an instance of #{other.class} to an instance of #{utc.class} is deprecated. This behavior will raise " \
```
Comment (jhawthorn) [inline activesupport/lib/active_support/time_with_zone.rb:308]: (non-blocking)

```suggestion
        "Adding an instance of #{other.class} to an instance of #{self.class} is deprecated. This behavior will raise " \
```

The user wasn't adding to `utc.class` (which is always Time) but `TimeWithZone`, I believe adding two `Time` together will already raise (which is how we get here).

File: activesupport/lib/active_support/time_with_zone.rb [hunk 7]
```diff
@@ -303,6 +303,13 @@ def +(other)
         result = utc + other
         result.in_time_zone(time_zone)
       end
+    rescue TypeError
```
Comment (jhawthorn) [inline activesupport/lib/active_support/time_with_zone.rb:306]: Having fairly generic rescues like this can be confusing if 

Could we make this rescue only capture only exceptions from the `utc + other` line? That way we're only going to rescue AT MOST exceptions we would have in previous versions of Rails.

**Review by jhawthorn — CHANGES_REQUESTED** [review]
Thanks for catching this. I left a few small suggestions

**Comment by jhawthorn** [comment]
Let's tackle this in a separate PR, but we should probably also deprecate/remove this in `Time#since`, which also has a rescue which was intended to switch to using `DateTime` to provide more resolution (which again is no longer possible and is just a mistake).

```
>> Time.now.since(Time.now)
=> Mon, 30 Jan 2079 09:13:28 -0700
>> _.class
=> DateTime
```

**Comment by Schwad** [comment]
> Thanks for catching this. I left a few small suggestions

Thank you for this. I will address the suggestions in the morning UK time and then squash the commits.

**Comment by Schwad** [comment]
> Let's tackle this in a separate PR, but we should probably also deprecate/remove this same behaviour in `Time#since`, which also has a rescue which was intended to switch to using `DateTime` to provide more resolution (which again is no longer possible and is just a mistake).
> 
> ```
> >> Time.now.since(Time.now)
> => Mon, 30 Jan 2079 09:13:28 -0700
> >> _.class
> => DateTime
> ```

I've taken [a stab at this](https://github.com/rails/rails/pull/52355). It does not currently have a CHANGELOG entry but I'm thinking it can be added to the CHANGELOG entry from this PR after it merges.

**Comment by jhawthorn** [comment]
Thank you! Merged in #52370
